### PR TITLE
Background QMP handling, tweak to machine definition and agent handling

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

File: lxd/container.go [hunk 1]
```diff
@@ -41,10 +41,9 @@ import (
 )
 
 func init() {
-	// Expose instanceLoadNodeAll to the device package converting the response to a slice of Instances.
```
Comment (tomponline) [inline lxd/container.go]: @stgraber did you intend to get ride of this comment line?

File: lxd/containers.go [hunk 2]
```diff
@@ -247,6 +248,21 @@ func containersRestart(s *state.State) error {
 	return nil
 }
 
+func vmMonitor(s *state.State) error {
```
Comment (tomponline) [inline lxd/containers.go]: @stgraber perhaps we could move this into a `vms.go` file as being in `containers.go` doesn't seem appropriate.

File: lxd/qmp/monitor.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+	return monitor, nil
+}
+
+func (m *Monitor) run() error {
+	// Start ringbuffer monitoring go routine.
+	go func() {
+		for {
+			// Read the ringbuffer.
+			resp, err := m.qmp.Run([]byte(`{"execute": "ringbuf-read", "arguments": {"device": "vserial", "size": 16, "format": "utf8"}}`))
```
Comment (tomponline) [inline lxd/qmp/monitor.go]: @stgraber is 16 bytes the max length of any of our ring-buffer messages?

File: lxd/vm_qemu.go [hunk 4]
```diff
@@ -314,6 +314,37 @@ func (vm *vmQemu) getStoragePool() (storagePools.Pool, error) {
 	return vm.storagePool, nil
 }
 
+func (vm *vmQemu) eventHandler() func(event string, data map[string]interface{}) {
```
Comment (tomponline) [inline lxd/vm_qemu.go]: @stgraber please could you call this monitorEventHandler() so its clear what sort of events is it handling, it also helps to show its related to the getMonitorPath() function. Perhaps for consistency `getMonitorPath` should be renamed to `monitorPath()` too, or `getMonitorEventHandler()` might actually be more appropriate because it itself is not the event handler, it returns an event handler function (this caught me out first pass).

File: lxd/vm_qemu.go [hunk 5]
```diff
@@ -557,13 +588,15 @@ func (vm *vmQemu) Start(stateful bool) error {
 	}
 
 	args := []string{
+		"-S",
```
Comment (tomponline) [inline lxd/vm_qemu.go]: @stgraber this delays the CPU start until a specific command is sent to the qmp, out of interest why is this needed? Thanks

File: lxd/daemon.go [hunk 6]
```diff
@@ -1017,6 +1018,9 @@ func (d *Daemon) Ready() error {
 	// Restore containers
 	containersRestart(s)
 
+	// Start monitoring VMs again
```
Comment (tomponline) [inline lxd/daemon.go]: @stgraber the "again" bit here caught my eye, what does it mean?

Comment (stgraber) [reply lxd/daemon.go]: It stopped when the LXD daemon shut down so has to be restarted for all running VMs on startup.

File: lxd/vm_qemu.go (hunk 5 above)
Comment (stgraber) [reply lxd/vm_qemu.go]: So there is no possible race that would cause us to miss some events.

File: lxd/vm_qemu.go (hunk 4 above)
Comment (stgraber) [reply lxd/vm_qemu.go]: Will do

File: lxd/qmp/monitor.go (hunk 3 above)
Comment (stgraber) [reply lxd/qmp/monitor.go]: Yes, the serial device is configured with a size of 16B.

Comment (tomponline) [reply lxd/qmp/monitor.go]: @stgraber thanks, I wonder if its worth defining a constant for that value and then using it here and down on line https://github.com/lxc/lxd/pull/6527/files#diff-8c876c3968dd86efda1f7fcd87e6c34aR1027 to show they are related and avoid any bugs where one is increased without the other.

File: lxd/container.go (hunk 1 above)
Comment (stgraber) [reply lxd/container.go]: Nope, will fix

File: lxd/containers.go (hunk 2 above)
Comment (stgraber) [reply lxd/containers.go]: I was expecting us to eventually rename this to instances.go at which point things will make more sense

**Comment by stgraber** [comment]
Updated based on feedback.

File: lxd/vm_qemu.go [hunk 7]
```diff
@@ -990,6 +1040,10 @@ driver = "virtio-scsi-pci"
 bus = "qemu_pcie1"
 addr = "0x0"
 
+# Graphics card
```
Comment (tomponline) [inline lxd/vm_qemu.go]: @stgraber was this required for the qmp work or just part of planned work to add GPU support?

Comment (stgraber) [reply lxd/vm_qemu.go]: No, it's just another tweak to our machine type I had around to allow for people to play with GUI installs. Combined with `raw.qemu` all they need is to pass `-vnc :1` to have a working graphical console.
